functions/go/apply-setters/set.go

```diff
@@ -0,0 +1,223 @@
+package main
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"sigs.k8s.io/kustomize/kyaml/errors"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+var _ kio.Filter = Set{}
+
+// Set sets resource field values with setter references
+type Set struct {
+	// Setters holds the user provided values for all the setters
+	Setters []Setter `json:"setters,omitempty" yaml:"setters,omitempty"`
+}
+
+type Setter struct {
+	// Name is the name of the setter
+	Name string `json:"name,omitempty" yaml:"name,omitempty"`
+
+	// Value is the input value for setter
+	Value string `json:"value,omitempty" yaml:"value,omitempty"`
+}
+
+const SetPrefix = "# kpt-set: "
+
+// Filter implements Set as a yaml.Filter
+func (s Set) Filter(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	for i := range nodes {
+		err := accept(&s, nodes[i])
+		if err != nil {
+			return nil, errors.Wrap(err)
+		}
+	}
+	return nodes, nil
+}
+
+// visitMapping
+func (s *Set) visitMapping(object *yaml.RNode, path string) error {
+	return object.VisitFields(func(node *yaml.MapNode) error {
+		if node == nil || node.Key == nil || node.Key.YNode() == nil ||
+			node.Value == nil || node.Value.YNode() == nil {
+			return nil
+		}
+		// the aim of this method is to apply-setter for sequence nodes
+		if node.Value.YNode().Kind != yaml.SequenceNode {
+			// return if it is not a sequence node
+			return nil
+		}
+
+		setterComment := node.Key.YNode().LineComment
+		if !strings.HasPrefix(setterComment, SetPrefix) {
+			return nil
+		}
+		setterComment = strings.TrimSpace(strings.TrimPrefix(setterComment, SetPrefix))
+		if !shouldSet(setterComment, s.Setters) {
+			// this means there is no intent from user to modify this setter tagged resources
+			return nil
+		}
+
+		// since this setter pattern is found on sequence node, make sure that it is
+		// not interpolation of setters, it should be simple setter e.g. ${environments}
+		if !validArraySetterPattern(setterComment) {
+			return errors.Errorf("setter interpolation is not allowed for sequence nodes: %q", setterComment)
```

Comment:
would a user understand what a "sequence node" mean (Hint: no) ? Refer to the style guide for best-practices for error messages. You use the phrase "array setter" further down.